### DISCUSSION TITLE
Incorrect conversion (rasterization process) of PDF file to magenta color scale

### DISCUSSION THREAD

**Discussion by MisiekDP** [ImageMagick/ImageMagick#5567] [Help]
When converting a PDF file to another PDF file with a white-magenta color scale, whole bars appear instead of the color scale.
Used command: convert -density 600 +level-colors magenta,white +antialias -background white -alpha remove -quality 100 input.pdf output.pdf

On the other hand, if I export the file as a JPG (I will rasterize it), the conversion is correct, but I would prefer to do it with one command without much quality loss.

Btw. 
maybe there is some other method that will allow me to just change the color scale to magenta (which is all I care about)?

1_Original_file:
<img width="702" alt="1_Original_file" src="https://user-images.githubusercontent.com/22171619/190451950-62f21ed9-b358-4865-a767-7ce8bf6da699.png">

2_Not_properly_converted_file (command: `convert -density 600 +level-colors magenta,white +antialias -background white -alpha remove -quality 100 input.pdf output.pdf`):
<img width="702" alt="2_Not_properly_converted_file" src="https://user-images.githubusercontent.com/22171619/190451983-a9ba0b23-6e01-478e-87d4-fcb0d3597a7c.png">

3_Properly_converted_file (how it should look) (PDF -> JPG (rasterize) -> JPG (convert to color scale) -> PDF):
<img width="702" alt="3_Properly_converted_file" src="https://user-images.githubusercontent.com/22171619/190452003-730d9ece-4949-4c80-865d-074b88484942.png">

This same problem on:
- Windows 10 19044.1826, Ghostscript 9.56.1, ImageMagick 7.1.0.47,
- RaspberryPi, Raspbian 11, Kernel 4.15.18-v7, Shell: bash 5.1.4, Ghostscript 9.53.3 (2020-10-01), ImageMagick 6.9.11-60 Q16 arm.

Example PDF file: 
[Cognito_Forms_DPA.pdf](https://github.com/ImageMagick/ImageMagick/files/9577547/Cognito_Forms_DPA.pdf)

**Comment by fmw42** [comment]
What version of Imagemagick and what platform/OS?

**Reply by fmw42** [reply]
Works fine for me on IM 6.9.12-63 on Mac OSX Monterey using Ghostscript 9.55. Please check your version of Ghostscript

gs --version

If not current, please upgrade and try again.

Also your syntax is not quite proper. Try the following (set the density, then read the input pdf, then do the processing as follows, then write the output).

`convert -density 600 B_jajecznica_kopia.pdf +level-colors magenta,white -background white -alpha background -alpha remove -quality 100 B_jajecznica_kopia_out.pdf
`

[B_jajecznica_kopia_out.pdf](https://github.com/ImageMagick/ImageMagick/files/9577064/B_jajecznica_kopia_out.pdf)

If on Imagemagick 7, use magick in place of convert.

**Reply by MisiekDP** [reply]
> What version of Imagemagick and what platform/OS?

I updated my first post.

**Reply by fmw42** [reply]
Did you try my command and change convert to magick?

`magick -density 600 B_jajecznica_kopia.pdf +level-colors magenta,white -background white -alpha background -alpha remove -quality 100 B_jajecznica_kopia_out.pdf`

**Reply by MisiekDP** [reply]
Yeah, I tried it and it doesn't works. The PDF file I uploaded is not correct. I tried to find one other file with this same problem but I don't have it now :/ I can't upload the original PDF file because it has a private data.

I tried the command on Windows without success: `>magick.exe -density 600 input.pdf +level-colors magenta,white -background white -alpha background -alpha remove -quality 100 output.pdf` :/

**Reply by fmw42** [reply]
One of the IM developers will need to check on Windows. Sorry, I do not use Windows.

**Reply by MisiekDP** [reply]
Ok @fmw42 I have a example file.

[Cognito_Forms_DPA.pdf](https://github.com/ImageMagick/ImageMagick/files/9577525/Cognito_Forms_DPA.pdf)

I got this problem when I use `>magick.exe -density 600 Cognito_Forms_DPA.pdf +level-colors magenta,white -background white -alpha background -alpha remove -quality 100 out.pdf`:

![image](https://user-images.githubusercontent.com/22171619/190469091-0a81ff44-f7f6-4577-a312-a94dd1d626c7.png)

Do you have this same problem maybe?

**Reply by fmw42** [reply]
Works fine for me with GS 9.55 and IM 7.1.0-48

`magick -density 600 Cognito_Forms_DPA.pdf +level-colors magenta,white -background white -alpha background -alpha remove -quality 100 Cognito_Forms_DPA_out.pdf`

[Cognito_Forms_DPA_out.pdf](https://github.com/ImageMagick/ImageMagick/files/9577687/Cognito_Forms_DPA_out.pdf)

**Reply by MisiekDP** [reply]
Hmm, it is really weird... I can check it on Raspbian too.

**Reply by MisiekDP** [reply]
Damn... On the Raspbian it works correctly too :/

Is it possible to download portable edition of previous ImageMagick version? I could check this same version on Windows.

**Comment by snibgo** [comment]
Your input PDF is mostly transparent. It is even transparent *black*. So you have black (opaque) text on a black (transparent) background.

The obvious suggestion is, immediately after reading the PDF, to flatten against white. *Then* mess around changing colours.

**Reply by MisiekDP** [reply]
Ok but how can I do this with ImageMagick? Is it possible to do with single command?

**Reply by MisiekDP** [reply]
And btw, the output is different of this file on Windows and different on Linux (Raspbian). On Raspbian I don't have this problem, now I see.
And there are different versions of GS and IM.

**Reply by fmw42** [reply]
See my command below

**Comment by fmw42** [comment]
Per snibgo's suggestion, try moving the +level-colors to after the transparent processing, near the end of the command

`magick -density 600 Cognito_Forms_DPA.pdf -background white -alpha background -alpha remove +level-colors magenta,white -quality 100 Cognito_Forms_DPA_out.pdf`

**Reply by MisiekDP** [reply]
OMG it works now on file which had a problem :D

**Reply by fmw42** [reply]
Thank snibgo. He caught the issue.

**Reply by MisiekDP** [reply]
Thank u guys, this is the support that I love 👌

**Comment by snibgo** [comment]
It is a common issue with PDF files. They often have opaque text on a transparent background.
